url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/Simulation.java

```diff
@@ -0,0 +1,1042 @@
+									(avgQPowerPerHourPerInst[counter]) * mcrunsRatio, 
+									DBResults.COL_INSTRESULTS_HOURLY_EN);
+							avgPPowerPerHourPerInst[counter] = 0;
+							avgQPowerPerHourPerInst[counter] = 0;
+							counter++;
+						}
+					}
+					mccount++;
+					//		  			percentage = (int)(0.75 * mccount * 100.0 / (mcruns * endTick));
+					//		  			System.out.println("Percentage: " + percentage + " - " + mccount);
+					//		  			objRun.put("percentage", 25 + percentage);
+					//		  	  		DBConn.getConn().getCollection(MongoRuns.COL_RUNS).update(query, objRun);
+				}
+				for(Installation installation: installations) {
+					installation.updateCost(tick); // update the rest of the energy
+					m.addKPIs(installation.getId(), 
+							installation.getMaxPower() * mcrunsRatio, 
+							installation.getAvgPower() * mcrunsRatio, 
+							installation.getEnergy() * mcrunsRatio, 
+							installation.getCost() * mcrunsRatio,
+							installation.getEnergy() * co2 * mcrunsRatio);
+					installation.addAppliancesKPIs(m, mcrunsRatio, co2);
+					installation.addActivitiesKPIs(m, mcrunsRatio, co2);
+				}
+				cost = totalInstCost();
+				m.addKPIs(DBResults.AGGR, 
+						maxPower * mcrunsRatio, 
+						avgPower * mcrunsRatio, 
+						energy * mcrunsRatio, 
+						cost * mcrunsRatio,
+						energy * co2 * mcrunsRatio);
+				if(i+1 != mcruns) setup(true);
+
+			}
+			// Write installation results to csv file
+			if (!resources_path.endsWith("/"))
+				resources_path += "/";
+			String filename = resources_path + dbname + ".csv";
+			//  			System.out.println(filename);
+			File csvFile = new File(filename);
+			FileWriter fw = new FileWriter(csvFile);
+			String row = "tick";
+			for(Installation installation: installations) {
+				row += "," + installation.getName() + "_p";
+				row += "," + installation.getName() + "_q";
+			}
+			fw.write(row+"\n");
+			for(int i = 0; i < endTick; i++) {
+				row = String.valueOf(i);
+				for(Installation installation: installations) {
+					if (useDerby)
+					{
+						ResultSet tickResult = ((DerbyResults)m).getTickResultForInstallation(i, 
+								installation.getId(),  
+								DBResults.COL_INSTRESULTS);
+						while (tickResult.next())
+						{
+							double p = tickResult.getDouble(3);
+							double q = tickResult.getDouble(4);
+							row += "," + p;
+							row += "," + q;
+						}	
+					}
+					else
+					{
+						DBObject tickResult =  ((MongoResults)m).getTickResultForInstallation(i, 
+								installation.getId(),  
+								DBResults.COL_INSTRESULTS);
+						double p = ((Double)tickResult.get("p")).doubleValue();
+						double q = ((Double)tickResult.get("q")).doubleValue();
+						row += "," + p;
+						row += "," + q;
+					}
+				}  				
+				fw.write(row+"\n");
+			}
+			fw.flush();
+			fw.close();
+			// End of file writing
+			System.out.println("Zipping output files...");
+			byte[] buffer = new byte[1024];
+			FileOutputStream fos = new FileOutputStream(filename + ".zip");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at lines 776, 778, 782, 788, 793. The resource is closed at lines 785, 796, 798, 799. There are other execution paths that do not contain closure statements, for example, when FileInputStream constructor throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).